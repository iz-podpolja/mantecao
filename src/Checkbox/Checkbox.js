import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { colors } from '../variables'
import { CheckIcon, BlankIcon } from '../icons'

const Container = styled.div`display: flex;`
const CheckboxDiv = styled.div`
  flex: 0 1 auto;
  margin-right: 1em;
`
const LabelDiv = styled.div`flex: 1 1 auto;`
const ChildrenDiv = styled.div`padding-top: 0.3em;`
const LabelText = styled.span`
  display: flex;
  cursor: ${props => (props.clickable ? 'pointer' : 'inherit')};
`
const CheckboxButton = styled.div`
  background: ${props => (props.checked ? props.color : colors.white)};
  border: 1px solid ${props => (props.checked ? props.color : 'currentColor')};
  border-radius: 2px;
  color: ${props => (props.checked ? colors.white : 'currentColor')};
  cursor: ${props => (props.disabled ? 'not-allowed' : 'pointer')};
  display: flex;
  opacity: ${props => (props.disabled ? '0.5' : '1')};
  padding: 0.1em;
`

class Checkbox extends Component {
  constructor(props) {
    super(props)
    this.state = {
      checked: props.checked
    }
  }
  componentWillReceiveProps(nextProps) {
    if (nextProps.checked !== undefined) {
      this.setState({ checked: nextProps.checked })
    }
  }
  render() {
    const {
      checked,
      disabled,
      label,
      color,
      children,
      allowLabelClick
    } = this.props
    return (
      <Container>
        <CheckboxDiv>
          <CheckboxButton
            color={color}
            disabled={disabled}
            checked={this.state.checked}
            onClick={this._toggleCheckbox}
          >
            {this.state.checked === true
              ? <CheckIcon
                  style={{
                    width: '1.2em',
                    height: '1.2em',
                    strokeWidth: '2.5px'
                  }}
                />
              : <BlankIcon style={{ width: '1.2em', height: '1.2em' }} />}
          </CheckboxButton>
        </CheckboxDiv>
        <LabelDiv>
          {label &&
            <LabelText
              clickable={allowLabelClick && !disabled}
              onClick={disabled ? () => false : this._toggleCheckbox}
            >
              {label}
            </LabelText>}
          {children && <ChildrenDiv>{children}</ChildrenDiv>}
        </LabelDiv>
      </Container>
    )
  }
  _toggleCheckbox = () => {
    const { onChange } = this.props
    if (!this.props.disabled) {
      this.setState({ checked: !this.state.checked }, () => {
        onChange(this.state.checked)
      })
    }
  }
}

Checkbox.defaultProps = {
  color: colors.blueBayoux
}

Checkbox.propTypes = {
  checked: PropTypes.bool,
  color: PropTypes.string,
  disabled: PropTypes.bool,
  label: PropTypes.string,
  allowLabelClick: PropTypes.bool,
  children: PropTypes.node,
  onChange: PropTypes.func.isRequired
}

export default Checkbox
