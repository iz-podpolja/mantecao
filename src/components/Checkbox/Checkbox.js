import React, {PropTypes, Component} from 'react'

import {colors, iconSizes} from '../../variables'
import {CheckIcon} from '../../icons'

const defaultContainerStyle = {
  display: 'flex'
}
const defaultLabelStyle = {
  display: 'block',
  fontWeight: 600,
  marginBottom: '0.5rem',
  maxWidth: '100%'
}
const defaultCheckboxStyle = {
  display: 'block',
  backgroundColor: colors.white,
  border: '1px solid',
  boxSizing: 'border-box',
  padding: '0.1rem',
  borderRadius: '2px',
  cursor: 'pointer',
  borderColor: 'currentColor',
  position: 'relative',
  height: '1.8rem',
  width: '1.8rem',
  margin: '0.2rem 0 0 0',
  fontSize: '0px', // fixes weird white-space issues
  userSelect: 'none' // fixes text select on multi-click
}
const checkedStyle = {
  color: colors.white,
  backgroundColor: colors.pacificBlue,
  borderColor: colors.pacificBlue
}
const disabledStyle = {
  color: colors.white,
  cursor: 'not-allowed',
  borderColor: colors.silver
}

/**
 * with label, description and hint
 */
class Checkbox extends Component {
  constructor (props) {
    super(props)
    this.state = {}
    // we need a state for this because it can be controlled or non-controlled
    // i.e. - given a 'checked' or 'defaultChecked' (or both - 'checked' wins)
    if (props.checked !== undefined) {
      this.state.checked = props.checked
    } else if (props.defaultChecked !== undefined) {
      this.state.checked = props.defaultChecked
    } else {
      this.state.checked = false
    }
  }
  componentWillReceiveProps (nextProps) {
    if (nextProps.checked !== undefined) {
      this.setState({checked: nextProps.checked})
    }
  }
  render () {
    const {error, warning, labelStyle, disabled, label, description, hint, style} = this.props

    let computerContainerStyle = Object.assign({}, defaultContainerStyle, style)
    let textColorStyle = Object.assign({},
      warning ? {color: colors.goldenTainoi} : {},
      error ? {color: colors.amaranth} : {}
    )
    let computedLabelStyle = Object.assign({}, defaultLabelStyle, labelStyle)
    let computedHintStyle = Object.assign({}, textColorStyle)
    let computedCheckboxStyle = Object.assign({},
      defaultCheckboxStyle,
      this.state.checked ? checkedStyle : {},
      disabled ? disabledStyle : {},
      this.state.checked && disabled ? {backgroundColor: colors.silver} : {}
    )

    return (
      <div style={computerContainerStyle}>
        <div style={{flex: '0 1 auto', marginRight: '0.7rem'}}>
          <label
            onClick={this._toggleCheckbox}
            htmlFor={this.refs.checkboxInput}
            style={computedCheckboxStyle}
          >
            {this.state.checked === true
              ? <CheckIcon style={{strokeWidth: '3.5px', width: '1.4rem', height: '1.4rem', verticalAlign: 'none'}} />
              : null
            }
            <input
              ref='checkboxInput'
              style={{visibility: 'hidden', position: 'absolute', left: 0}}
              type='checkbox'
              disabled={disabled}
              checked={this.state.checked}
              onChange={this._onChange}
            />
          </label>
        </div>
        <div style={{flex: '1 1 auto'}}>
          {label
            ? <label style={computedLabelStyle}>{label}</label>
            : null
          }
          {description
            ? <div style={{marginBottom: '0.5rem'}}>{description}</div>
            : null
          }
          {hint
            ? <div style={computedHintStyle}>{hint}</div>
            : null
          }
        </div>
      </div>
    )
  }
  _toggleCheckbox = (e) => {
    e.preventDefault()
    if (!this.props.disabled) {
      this.setState(
        {checked: !this.state.checked},
        () => {
          this.props.onChange(this.state.checked)
        }
      )
    }
  }
  _onChange = (e) => {
    // with a controlled input, react requires an onchange handler - do nothing
  }
}

Checkbox.propTypes = {
  checked: PropTypes.bool,
  defaultChecked: PropTypes.bool,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  hint: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  label: PropTypes.string,
  labelStyle: PropTypes.object,
  onChange: PropTypes.func.isRequired,
  warning: PropTypes.bool
}

export default Checkbox
