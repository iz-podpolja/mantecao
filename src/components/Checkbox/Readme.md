Example:

    <Checkbox
      onChange={(e)=>console.log(e.target.checked)}
      label="hello checkbox"
      description="hello checkbox description"
      checked={true}
      error
      hint="maybe there is an error"
    />
