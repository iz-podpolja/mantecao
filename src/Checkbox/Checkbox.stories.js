import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import Checkbox from './Checkbox'
import InlineAlert from '../InlineAlert'

storiesOf('Checkbox', module)
  .add('default', () =>
    <Checkbox
      label="hello checkbox"
      onChange={action('checkbox clicked')}
      checked={true}
    />
  )
  .add('with children', () =>
    <Checkbox
      label="hello checkbox"
      onChange={action('checkbox clicked')}
      checked={true}
    >
      <InlineAlert type="error">You can send errors in as children</InlineAlert>
    </Checkbox>
  )
  .add('disabled', () =>
    <Checkbox
      label="hello checkbox"
      disabled
      onChange={action('checkbox clicked')}
      checked={true}
    />
  )
  .add('with label clickable', () =>
    <Checkbox
      label="hello checkbox"
      allowLabelClick
      onChange={action('checkbox clicked')}
      checked={true}
    />
  )
