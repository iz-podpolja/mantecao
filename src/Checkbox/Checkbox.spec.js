import React from 'react'
import renderer from 'react-test-renderer'
import Checkbox from './Checkbox'

describe('Checkbox', () => {
  it('renders with a label', () => {
    const tree = renderer
      .create(
        <Checkbox
          label="hello checkbox"
          onChange={() => false}
          checked={true}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders without a label', () => {
    const tree = renderer
      .create(<Checkbox onChange={() => false} checked={true} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders checked', () => {
    const tree = renderer
      .create(
        <Checkbox
          label="hello checkbox"
          onChange={() => false}
          checked={true}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders unchecked', () => {
    const tree = renderer
      .create(
        <Checkbox
          label="hello checkbox"
          onChange={() => false}
          checked={false}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('changes state with a click', () => {})
})
