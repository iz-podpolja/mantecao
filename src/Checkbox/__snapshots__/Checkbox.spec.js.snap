// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders checked 1`] = `
<div
  className="sc-bdVaJa gzoWJL"
>
  <div
    className="sc-bwzfXH fThnKs"
  >
    <div
      checked={true}
      className="sc-EHOje fJaLOh"
      color="#556285"
      disabled={undefined}
      onClick={[Function]}
    >
      <svg
        className={undefined}
        style={
          Object {
            "display": "inline-block",
            "fill": "none",
            "height": "1.2em",
            "pointerEvents": "none",
            "stroke": "currentColor",
            "strokeWidth": "2.5px",
            "verticalAlign": "middle",
            "width": "1.2em",
          }
        }
        viewBox="0 0 32 32"
      >
        <g
          transform="translate(0, -2)"
        >
          <path
            d="M3 18.2L11.4 26.7C11.8 27.1 12.4 27.1 12.8 26.7L30.1 9"
          />
        </g>
      </svg>
    </div>
  </div>
  <div
    className="sc-htpNat kruefc"
  >
    <span
      className="sc-ifAKCX gEjVhK"
      onClick={[Function]}
    >
      hello checkbox
    </span>
  </div>
</div>
`;

exports[`Checkbox renders unchecked 1`] = `
<div
  className="sc-bdVaJa gzoWJL"
>
  <div
    className="sc-bwzfXH fThnKs"
  >
    <div
      checked={false}
      className="sc-EHOje byiWrQ"
      color="#556285"
      disabled={undefined}
      onClick={[Function]}
    >
      <svg
        className={undefined}
        style={
          Object {
            "display": "inline-block",
            "fill": "none",
            "height": "1.2em",
            "pointerEvents": "none",
            "stroke": "currentColor",
            "strokeWidth": 1.5,
            "verticalAlign": "middle",
            "width": "1.2em",
          }
        }
        viewBox="0 0 32 32"
      >
        <g
          transform={undefined}
        />
      </svg>
    </div>
  </div>
  <div
    className="sc-htpNat kruefc"
  >
    <span
      className="sc-ifAKCX gEjVhK"
      onClick={[Function]}
    >
      hello checkbox
    </span>
  </div>
</div>
`;

exports[`Checkbox renders with a label 1`] = `
<div
  className="sc-bdVaJa gzoWJL"
>
  <div
    className="sc-bwzfXH fThnKs"
  >
    <div
      checked={true}
      className="sc-EHOje fJaLOh"
      color="#556285"
      disabled={undefined}
      onClick={[Function]}
    >
      <svg
        className={undefined}
        style={
          Object {
            "display": "inline-block",
            "fill": "none",
            "height": "1.2em",
            "pointerEvents": "none",
            "stroke": "currentColor",
            "strokeWidth": "2.5px",
            "verticalAlign": "middle",
            "width": "1.2em",
          }
        }
        viewBox="0 0 32 32"
      >
        <g
          transform="translate(0, -2)"
        >
          <path
            d="M3 18.2L11.4 26.7C11.8 27.1 12.4 27.1 12.8 26.7L30.1 9"
          />
        </g>
      </svg>
    </div>
  </div>
  <div
    className="sc-htpNat kruefc"
  >
    <span
      className="sc-ifAKCX gEjVhK"
      onClick={[Function]}
    >
      hello checkbox
    </span>
  </div>
</div>
`;

exports[`Checkbox renders without a label 1`] = `
<div
  className="sc-bdVaJa gzoWJL"
>
  <div
    className="sc-bwzfXH fThnKs"
  >
    <div
      checked={true}
      className="sc-EHOje fJaLOh"
      color="#556285"
      disabled={undefined}
      onClick={[Function]}
    >
      <svg
        className={undefined}
        style={
          Object {
            "display": "inline-block",
            "fill": "none",
            "height": "1.2em",
            "pointerEvents": "none",
            "stroke": "currentColor",
            "strokeWidth": "2.5px",
            "verticalAlign": "middle",
            "width": "1.2em",
          }
        }
        viewBox="0 0 32 32"
      >
        <g
          transform="translate(0, -2)"
        >
          <path
            d="M3 18.2L11.4 26.7C11.8 27.1 12.4 27.1 12.8 26.7L30.1 9"
          />
        </g>
      </svg>
    </div>
  </div>
  <div
    className="sc-htpNat kruefc"
  />
</div>
`;
